aten/src/ATen/native/TensorIndexing.h

```diff
@@ -77,36 +125,454 @@ struct CAFFE2_API TensorIndex final {
 
   // Case 5: Slice represented in `{start, stop, step}` form,
   // where `start` / `stop` / `step` can be integer or `at::indexing::None`
-  TensorIndex(std::initializer_list<c10::optional<int64_t>> init_list);
+  TensorIndex(std::initializer_list<c10::optional<int64_t>> slice) : type_(TensorIndexType::Slice) {
+    if (slice.size() == 0) {
+      slice_ = impl::unpackSlice(c10::nullopt, c10::nullopt, c10::nullopt);
+    } else if (slice.size() == 2) {
+      slice_ = impl::unpackSlice(*slice.begin(), *(slice.begin() + 1), c10::nullopt);
+    } else if (slice.size() == 3) {
+      slice_ = impl::unpackSlice(*slice.begin(), *(slice.begin() + 1), *(slice.begin() + 2));
+    } else {
+      TORCH_CHECK_VALUE(
+        false,
+        "Expected 0 / 2 / 3 elements in the braced-init-list to represent a slice index, but got ",
+        slice.size(),
+        " element(s)");
+    }
+  }
 
   // Case 5: Tensor value
-  TensorIndex(Tensor tensor);
+  TensorIndex(Tensor tensor) : tensor_(tensor), type_(TensorIndexType::Tensor) {}
+
+  inline bool is_none() const {
+    return type_ == TensorIndexType::None;
+  }
+
+  inline bool is_ellipsis() const {
+    return type_ == TensorIndexType::Ellipsis;
+  }
 
-  bool is_none() const;
-  bool is_ellipsis() const;
+  inline bool is_integer() const {
+    return type_ == TensorIndexType::Integer;
+  }
 
-  bool is_integer() const;
-  int64_t integer() const;
+  inline int64_t integer() const {
+    return integer_;
+  }
 
-  bool is_boolean() const;
-  bool boolean() const;
+  inline bool is_boolean() const {
+    return type_ == TensorIndexType::Boolean;
+  }
 
-  bool is_slice() const;
-  const Slice& slice() const;
+  inline bool boolean() const {
+    return boolean_;
+  }
 
-  bool is_tensor() const;
-  const Tensor& tensor() const;
+  inline bool is_slice() const {
+    return type_ == TensorIndexType::Slice;
+  }
+
+  inline const impl::Slice& slice() const {
+    return slice_;
+  }
+
+  inline bool is_tensor() const {
+    return type_ == TensorIndexType::Tensor;
+  }
+
+  inline const Tensor& tensor() const {
+    return tensor_;
+  }
 
  private:
   int64_t integer_;
   bool boolean_;
-  Slice slice_;
+  impl::Slice slice_;
   Tensor tensor_;
   TensorIndexType type_;
 };
 
 CAFFE2_API std::ostream& operator<<(std::ostream& stream, const TensorIndex& tensor_index);
 CAFFE2_API std::ostream& operator<<(std::ostream& stream, const std::vector<TensorIndex>& tensor_indices);
 
+namespace impl {
+static inline Tensor applySlice(
+    const Tensor& self,
+    int64_t dim,
+    int64_t start,
+    int64_t stop,
+    int64_t step,
+    bool ensure_view,
+    bool is_tracing,
+    const at::Device& self_device,
+    const IntArrayRef& self_sizes) {
+  // TODO: implement negative step
+  TORCH_CHECK_VALUE(step > 0, "step must be greater than zero");
```

Comment:
It kind of looks like this error test might be redundant with whatever error test `self.slice` does.

Reply:
(No action needed on this PR, I assume this was faithfully translated from the original).

Reply:
I guess the idea is to give a better error message than what the underlying implementation would have given.